File: 113_Path_Sum_II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self, node, remainingSum, pathNodes, pathsList):
        
        if not node:
            return 
        
        pathNodes.append(node.val)
        
    
        if remainingSum == node.val and not node.left and not node.right:
            pathsList.append(list(pathNodes))
        else:    
            self.helper(node.left, remainingSum - node.val, pathNodes, pathsList)
            self.helper(node.right, remainingSum - node.val, pathNodes, pathsList)
            

        pathNodes.pop()    
    
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        pathsList = []
        self.helper(root, sum, [], pathsList)
        return pathsList
        

File: 113_Path_Sum_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void solve(TreeNode* root, int targetSum, int currSum, vector<int> vec, vector<vector<int>> &ans){    
        if(root==NULL) return;
        currSum+=root->val;
        vec.push_back(root->val);
        if(currSum==targetSum){
            if(!root->left and !root->right)
            ans.push_back(vec);
        }        
        solve(root->left,targetSum,currSum,vec,ans);
        solve(root->right,targetSum,currSum,vec,ans);   
    }
    
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> ans;
        if(root==NULL) return ans;
        vector<int> vec;
        int currSum=0;
        solve(root,targetSum,currSum,vec,ans);
        return ans;
    }
};